### ISSUE TITLE
Allow the target FPS of `display-*` configurable

### ISSUE THREAD

**Issue by hooke007** [mpv-player/mpv#11122]
Currently set `video-sync=display-resample` would make mpv always try to match the monitor's highest refresh rate. For those 240~360Hz monitors, enabling this option would introduce much frame-drop/delay.

### Expected behavior of the wanted feature

Add the new option like `video-sync-target-fps=xxx` to avoid this trouble.

### Alternative behavior of the wanted feature

N/A

### Log file

N/A

**Comment by christoph-heinrich** [comment]
The increased overhead comes from interpolation and not from `video-sync=display-*`, but reducing the targeted frame rate would certainly help with resource usage.

Instead of targeting a specific refresh rate, it would make more sense to cap it to a max so that it can still line up with lower refresh rates in mixed refresh rate systems.

But even that isn't an optimal solution, because when e.g. a cap of 80fps is set and the window is on a 120hz monitor, then targeting 60fps instead of capping out on 80fps would probably result in smoother playback.

Now with that being said, is there even a need for interpolation on such monitors? I'd imagine that the refresh rate is high enough that slightly imperfectly timed frames would be hardly noticeable. I have never used such a monitor though, so it's hard to say.

**Comment by hooke007** [comment]
> The increased overhead comes from interpolation

I didn't enable interpolation with display-resample. Enabling interpolation would increase more GPU consumption.

> But even that isn't an optimal solution, because when e.g. a cap of 80fps is set and the window is on a 120hz monitor, then targeting 60fps instead of capping out on 80fps would probably result in smoother playback.

This could ref some games, they could use fixed 1/2 or 1/4 refreash rate as vsync target.

**Comment by Dudemanguy** [comment]
Should `--override-display-fps` work for this?

**Comment by hooke007** [comment]
> Shouldn't --override-display-fps work for this?

Unfortunately, it made more dropped frames.

**Comment by Dudemanguy** [comment]
Can you post a log with that option set to 120 really quick?

**Comment by hooke007** [comment]
[fps.log](https://github.com/mpv-player/mpv/files/10387552/fps.log)

**Comment by Dudemanguy** [comment]
>[   0.104][v][vo/gpu] Assuming 120.000000 FPS for display sync.

So yes, the target FPS is actually configurable and a quick look at the code confirms this. However, when using a `display-*` video sync mode, this leads to worse results. The reason is because those modes are timed directly to vsync. Since your monitor has such a high refresh rate, the vsync block returns twice as fast (i.e. at a 240hz rate bit 120) which proceeds to mess up mpv's timing code. So the feature you want technically exists (`override-display-fbs`), but it's at odds with how `display-*` is designed so it can't possibly work well. I'll alter the title a bit for clarity for clarity.

As for implementing something like this, it wouldn't be impossible but it would take some work and careful thinking.

**Comment by Dudemanguy** [comment]
That sounds like a bug.

**Comment by BhaturaGuy** [comment]
Maybe it is.
I found 2 fixes for these, (Nvidia  GPU)
1. Turning off G-sync/freesync. (Not reliable)
2. Setting Monitor Technology to Fixed Refresh Rate for mpv in Nvidia Control Pannel. (Con: Black screen for 2 secs whenever mpv is opened, closed and fullscreen in and out.)

Can't reproduce my issue on AMD gpu.

**Comment by hooke007** [comment]
G-sync is uncompatiable with mpv. It's another known issue.

**Comment by BhaturaGuy** [comment]
Yeah, maybe I should open another issue for my problem, plus I have some questions too.

**Comment by dontpokethebear3893** [comment]
+1 for this feature request. I am currently toying with smooth motion and refresh rate handling on my 175hz monitor with gsync and this would solve most of my problems.

**Comment by snylonue** [comment]
> Now with that being said, is there even a need for interpolation on such monitors? I'd imagine that the refresh rate is high enough that slightly imperfectly timed frames would be hardly noticeable. I have never used such a monitor though, so it's hard to say.
Actually, the differences between interpolation on and off are noticeable on a 240Hz monitor.

Is this issue related to [the drawback described in wiki](https://github.com/mpv-player/mpv/wiki/Display-synchronization#display-sync)?
> Finally, --video-sync=display-* currently comes with one important drawback: Due to OpenGL's rather severe limitations when it comes to timing, the only way to reliably figure out when vsyncs happen is to actually draw a frame on every vsync. The consequence of this is that, even for 24 Hz video, you need to draw frames at 60 Hz even if they are the same frame over and over again - thus increasing power usage by a factor of 2x-3x in such a case. This is usually somewhat alleviated by caching the frame result, and redrawing the cached frame, instead of scaling and rendering the source frame every time. The problem that there's less time to render subtitles on a 60 Hz screen remains.

**Comment by hooke007** [comment]
@Dudemanguy Could you consider adding it into https://github.com/mpv-player/mpv/milestone/3 ?

> Since your monitor has such a high refresh rate, the vsync block returns twice as fast (i.e. at a 240hz rate instead of 120hz) which proceeds to mess up mpv's timing code which is being told that the display FPS is 120. 

Unfortunately, I noticed obvious frame-drops in my current display(4k95hz). I cannot even smoothly playback a 720p24hz short video. This sounds really ridiculous.
Which means large numbers of modern monitor would suffer this pain.

**Comment by Dudemanguy** [comment]
A framerate limiter is a good feature in general but out of scope for the upcoming release.

**Comment by kasper93** [comment]
> A framerate limiter is a good feature in general but out of scope for the upcoming release. We can revisit this after 0.37 I think.

While this can be added, the limiter is directly contradict the premise of **display**-sync. The main problem is that we cannot schedule vsyncs accurate enough in software to make perfectly smooth playback, without any strutter. That's why display-sync exist to rely on implicit vsync to sync frames correctly.

So actually the frame limiter will not improve situation, because it will introduce stutter in itself. The proper way to use display-sync is to set your display to certain refresh rate and let mpv sync to that.

Of course frame limiter is viable option, but like I said it will not do what you want it to do in fact.

**Comment by Jules-A** [comment]
> While this can be added, the limiter is directly contradict the premise of **display**-sync. The main problem is that we cannot schedule vsyncs accurate enough in software to make perfectly smooth playback, without any strutter. That's why display-sync exist to rely on implicit vsync to sync frames correctly.

That's not entirely true, with Freesync it's possible but Freesync only works properly in fullscreen. Currently you need to use a vf to multiply the video's FPS though which isn't really ideal.

**Comment by kasper93** [comment]
> That's not entirely true, with Freesync it's possible but Freesync only works properly in fullscreen.

I was not asking. Software syncing will always produce frame pacing issues. The fact VRR might eat them is beside the point. The important part is that the frame pacing would already by not smooth.

**Comment by bsolar17** [comment]
I think I'm encountering a similar issue on MacOS on a MacBook Pro with 120Hz monitor. The issue is very apparent when using Vulkan, but not with libmpv.

With this command, a 4K video is pretty jerky, showing a lot of Mistimed/Delayed frames happening during playback:

`mpv --no-config -vo=gpu-next -gpu-context=macvk -fs -video-sync=display-resample test.mp4`

With this command the same video is smooth:

`mpv --no-config -vo=gpu-next -gpu-context=macvk -fs test.mp4`

And with this one the video is also smooth, with only a handful of Mistimed/Delayed frames happening at the very beginning:

`mpv --no-config -vo=libmpv -video-sync=display-resample -fs test.mp4`

**Comment by Akemi** [comment]
@bsolar17 try with different options of `--macos-render-timer`, eg `precise` and `system`.

**Comment by bsolar17** [comment]
@Akemi, I tried with both options but they did not improve the jerkiness.

**Comment by snylonue** [comment]
> > This could ref some games, they could use fixed 1/2 or 1/4 refreash rate as vsync target.
> 
> This is already possible with `--opengl-swapinterval`/`--d3d11-sync-interval` and `--display-fps-override`. You need to manually calculate the display fps after being divided by the swap interval, but this works for me.


but such options won't solve frame the dropping issue as pointed out before.

**Comment by xCuri0** [comment]
Not sure if this is the right place for VRR jitter issues but I still have them using
```
vo=gpu-next
gpu-context=d3d11
hwdec=d3d11va
d3d11-sync-interval=0
```
On Windows 10 22H2 RX 580, I measured the frame times using CapFrameX (or any similar tool) and they are still unstable.

![Image](https://github.com/user-attachments/assets/a3115d2e-b891-4fac-80ad-f569106ab4ce)

The only players I found that display VRR properly are the built in Microsoft UWP ones with zero measurable jitter by monitor OSD which updates every frame. 

I guess Microsoft must be using a fixed timer or something to present frames because their VRR doesn't support variable rate video.

MPC-VR is a close second, but it seems to use milliseconds or windows timers internally so it has frame variations at a repeating pattern +/- 1ms.

**Comment by hooke007** [comment]
Repeat again this issue is not discussing about VRR.
I received the feedback that it doesn't happen on Windows. But I cannot reproduce for I've been back to using a 60Hz monitor for over a year now.
